Redefine fromAsync as “built-in async function”

js-choi opened the pull request:
This tweaks the spec to use the new “built-in async function” machinery in tc39/ecma262#2942. It resolves #43. See also #31, #36.

* It uses the same “This async method performs the following steps when called” language as proposal-async-iterator-helpers.
* It replaces returning Completion Records with direct returning of array values, which are now implicitly wrapped in Completion Records by the definition of built-in async function objects.
* It removes the wrapping async Abstract Closure, which is now implicitly defined by the definition of built-in async function objects.
* It tweaks the ending note to match the language in [Array.from’s note](https://tc39.es/ecma262/multipage/indexed-collections.html#:~:text=This%20method%20is%20an%20intentionally%20generic%20factory%20method;).
* It [tweaks the spec location of IfAbruptCloseAsyncIterator](https://github.com/tc39/proposal-array-from-async/pull/50#issuecomment-2844028670) to be “Operations on Iterator Objects” of “Abstract Operations”, to match IfAbruptCloseIterator. This is an editorial change.
* It tweaks IfAbruptCloseAsyncIterator to assert that its given value is a Completion Record. This matches [IfAbruptCloseAsyncIterator](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-ifabruptcloseiterator). This is an editorial change.
* It adds an `aoid` attribute for IfAbruptCloseAsyncIterator. This is an editorial change.
* IfAbruptCloseAsyncIterator now directly uses the return value of AsyncIteratorClose. This is an editorial change.
* Editorial formatting changes have been done, e.g., avoiding unnecessary HTML entities and making if/else subclauses use blocks if any alternative branch uses blocks.
* It lastly updates ecmarkup to v21.2.0 and ecma262-biblio to v2.1.2869.

This is the last step before creating a pull request to tc39/ecma262 <ins>(along with resolving the merge conflicts of tc39/ecma262#2942)</ins>. We’re almost there.

@nicolo-ribaudo, @ljharb, since you reviewed it before for Stage 3 (#14), would you mind reviewing before we merge this?

js-choi committed:
Spec: First pass at conforming to "built-in async method" system

js-choi committed:
Build: Upgrade ecmarkup to v21.2.0

ljharb commented on package.json:
```diff
@@ -26,7 +26,7 @@
   },
   "devDependencies": {
     "@tc39/ecma262-biblio": "2.1.2462",
```
you should probably update the biblio to the latest also

js-choi committed:
build: update ecma262-biblio to v2.1.2869

js-choi committed:
editorial: change IfAbruptCloseAsyncIterator location

The “Iterator Abstract Operations” subsection of “Iteration” from the
old version of proposal-iterator-helpers and
proposal-async-iterator-helpers is obsolete.

This means that the new IfAbruptCloseAsyncIterator operation (shared with proposal-async-iterator-helpers) no longer has a place to
live.

The ecma262 spec placed
proposal-iterator-helpers’ analogous IfAbruptCloseIterator operation
into a “Operations on Iterator Objects” subsection of “Abstract Operations”.

This commit places IfAbruptCloseAsyncIterator into an analogous
“Operations on Async Iterator Objects” subsection…although Async
Iterator Objects do not yet exist!

The editors may choose to name this new subsection whatever they want.
Or they might elect to not make a new subsection and foist
IfAbruptCloseAsyncIterator on the existing “Operations on Iterator
Objects” subsection instead, until proposal-async-iterator-helpers makes
Stage 4.

This is an editorial change.

js-choi commented:
FYI, I pushed an editorial change to the new IfAbruptCloseAsyncIterator operation’s location.

The “Iterator Abstract Operations” subsection of “Iteration” from the old version of proposal-iterator-helpers and proposal-async-iterator-helpers is obsolete.

This means that the new IfAbruptCloseAsyncIterator operation (from #25, shared with proposal-async-iterator-helpers) no longer has a place in the spec to live.

The ecma262 spec placed proposal-iterator-helpers’ analogous [IfAbruptCloseIterator](https://tc39.es/ecma262/#sec-ifabruptcloseiterator) operation into a [“Operations on Iterator Objects”](https://tc39.es/ecma262/#sec-operations-on-iterator-objects) subsection of “Abstract Operations”.

This commit places IfAbruptCloseAsyncIterator into an analogous new “Operations on Async Iterator Objects” subsection…although Async Iterator Objects do not yet exist (coming in proposal-async-iterator-helpers)!

The spec editors may choose to name this new subsection whatever they want. Or they might elect to not make a new subsection and foist IfAbruptCloseAsyncIterator on the existing “Operations on Iterator Objects” subsection instead, until proposal-async-iterator-helpers makes Stage 4.

Since this is a merely editorial tweak to the pull request, I don’t think the tweak deserves scrutiny until I make a pull request on tc39/ecma262#2942.

js-choi committed:
editorial: Change “Operations on Async Iterator Objects” to “Operations on Iterator Objects”

This matches the spec’s use of “Operations on Iterator Objects” to
contain async iterable operations AsyncIteratorClose.

js-choi committed:
editorial: add an assert to IfAbruptCloseAsyncIterator

This matches ecma262’s IfAbruptCloseIterator.

js-choi committed:
editorial: Remove year from title

js-choi committed:
editorial: Add aoid attribute for IfAbruptCloseAsyncIterator

js-choi committed:
Editorial changes to match Array.from spec, esformat

* IfAbruptCloseAsyncIterator now directly uses the return value of   AsyncIteratorClose. This is an editorial change.
* Editorial formatting changes have been done, e.g., avoiding unnecessary HTML entities and making if/else subclauses use blocks if any alternative branch uses blocks.

nicolo-ribaudo reviewed (approved):
Sorry for the delay!
